File: src/medium/BestTimeBuySellStockII.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buyIndex = 0
        sellIndex = None
        profit = 0

        arraySize = len(prices)

        for i in range(1, arraySize):
            if sellIndex is None: # we should buy
                if prices[i] < prices[buyIndex]:
                    buyIndex = i
                else:
                    sellIndex = i
                    # print('buy in', buyIndex)
            else: # we should sell
                if prices[i] > prices[sellIndex]:
                    sellIndex = i
                else:
                    # print('sell in', sellIndex)
                    profit += prices[sellIndex] - prices[buyIndex]
                    buyIndex = i
                    sellIndex = None

        if sellIndex is not None:
            profit += prices[sellIndex] - prices[buyIndex]

        return profit

File: src/medium/BestTimeBuySellStockII.go
package main

import "fmt"

func maxProfitBuySellStock(prices []int) int {
	localMin := prices[0]
	var profit int
	for i := 1; i < len(prices); i++ {
		if prices[i] < prices[i-1] {
			profit += prices[i-1] - localMin
			localMin = prices[i]
		}
	}
	profit += prices[len(prices)-1] - localMin
	return profit
}

func main() {
	fmt.Println(maxProfitBuySellStock([]int{7, 1, 5, 3, 6, 4}))
	fmt.Println(maxProfitBuySellStock([]int{1, 2, 3, 4, 5}))
	fmt.Println(maxProfitBuySellStock([]int{7, 6, 4, 3, 1}))
}
